dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/preheat/DefaultPreheatService.java

```diff
@@ -189,137 +161,97 @@ public Preheat preheat( PreheatParams params )
-                            List<? extends IdentifiableObject> objects = queryService.query( query );
-                            preheat.put( PreheatIdentifier.UID, objects );
-                        }
-                    }
-                }
-            }
-
-            if ( codeMap != null && (PreheatIdentifier.CODE == params.getPreheatIdentifier()
-                || PreheatIdentifier.AUTO == params.getPreheatIdentifier()) )
+            for ( Class<? extends IdentifiableObject> klass : uidMap.keySet() )
             {
-                for ( Class<? extends IdentifiableObject> klass : codeMap.keySet() )
-                {
-                    List<List<String>> identifiers = Lists.partition( Lists.newArrayList( codeMap.get( klass ) ),
-                        20000 );
-
-                    if ( !identifiers.isEmpty() )
-                    {
-                        for ( List<String> ids : identifiers )
-                        {
-                            Query query = Query.from( schemaService.getDynamicSchema( klass ) );
-                            query.setUser( preheat.getUser() );
-                            query.add( Restrictions.in( "code", ids ) );
-                            List<? extends IdentifiableObject> objects = queryService.query( query );
-                            preheat.put( PreheatIdentifier.CODE, objects );
-                        }
-                    }
-                }
+                List<List<String>> identifiers = Lists.partition( Lists.newArrayList( uidMap.get( klass ) ),
+                    20000 );
 
-                if ( uidMap.containsKey( User.class ) && !uidMap.get( User.class ).isEmpty() )
+                if ( !identifiers.isEmpty() )
                 {
-                    List<List<String>> identifiers = Lists.partition( Lists.newArrayList( uidMap.get( User.class ) ),
-                        20000 );
-
                     for ( List<String> ids : identifiers )
                     {
-                        Query query = Query.from( schemaService.getDynamicSchema( User.class ) );
+                        Query query = Query.from( schemaService.getDynamicSchema( klass ) );
                         query.setUser( preheat.getUser() );
                         query.add( Restrictions.in( "id", ids ) );
                         List<? extends IdentifiableObject> objects = queryService.query( query );
                         preheat.put( PreheatIdentifier.UID, objects );
                     }
                 }
+            }
+        }
 
-                if ( uidMap.containsKey( UserAuthorityGroup.class )
-                    && !uidMap.get( UserAuthorityGroup.class ).isEmpty() )
-                {
-                    List<List<String>> identifiers = Lists
-                        .partition( Lists.newArrayList( uidMap.get( UserAuthorityGroup.class ) ), 20000 );
+        if ( codeMap != null && (PreheatIdentifier.CODE == params.getPreheatIdentifier()) )
```

Comment:
You say only `REFERENCE` is supported and all others are treated as `REFERENCE` but here is a condition that tests for `CODE`. That confuses me.

Reply:
Hi @jbee, so `REFERENCE` is for `PreheatMode` and `ID` and `UID` is for `PreheatIdentifier` it will all be clear on next week metadata call.